File: web/src/lib/io/uri-resolver.ts
/**
 * Scheme-keyed URI resolver for asset loading on the web.
 *
 * Mirrors the C++ `cedar::UriResolver`. Handlers register for a scheme
 * ("https", "github", "bundled", "blob", "idb"); `load(uri)` parses the
 * scheme and dispatches.
 *
 * The native side (cedar) handles file://, http://, https://, github:,
 * bundled:// itself. On the web, http/https/github fetch via the browser;
 * file:// is unavailable; bundled:// looks up a build-time table.
 */
import { FileLoadError } from './errors';
import type { LoadOptions, LoadResult } from './file-loader';

export interface UriHandler {
	/** Lowercase scheme identifier without trailing punctuation. */
	scheme: string;
	load(uri: string, options: LoadOptions): Promise<LoadResult>;
}

const BLOB_PREFIX = 'blob:nkido:';

function extractScheme(uri: string): string {
	if (!uri) return '';
	// Bare paths and Windows drive letters are not URIs.
	if (uri.startsWith('/') || uri.startsWith('.')) return '';
	if (/^[a-zA-Z]:[\\/]/.test(uri)) return '';
	const colon = uri.indexOf(':');
	if (colon <= 0) return '';
	return uri.slice(0, colon).toLowerCase();
}

class UriResolver {
	private handlers = new Map<string, UriHandler>();
	private blobs = new Map<string, File | ArrayBuffer>();
	private inFlight = new Map<string, Promise<LoadResult>>();

	register(handler: UriHandler): void {
		this.handlers.set(handler.scheme.toLowerCase(), handler);
	}

	handlerFor(scheme: string): UriHandler | undefined {
		return this.handlers.get(scheme.toLowerCase());
	}

	async load(uri: string, options: LoadOptions = {}): Promise<LoadResult> {
		if (!uri) {
			throw new FileLoadError('invalid_format', 'empty URI');
		}

		// Concurrent loads of the same URI share the same network request.
		// Skipped when caller passes onProgress (progress reports would
		// interleave between callers in a confusing way).
		const dedupKey = options.signal || options.onProgress ? null : uri;
		if (dedupKey) {
			const existing = this.inFlight.get(dedupKey);
			if (existing) return existing;
		}

		const scheme = extractScheme(uri) || 'file';
		const handler = this.handlerFor(scheme);
		if (!handler) {
			throw new FileLoadError(
				'invalid_format',
				`no handler registered for scheme '${scheme}'`
			);
		}

		const promise = handler.load(uri, options).finally(() => {
			if (dedupKey) this.inFlight.delete(dedupKey);
		});
		if (dedupKey) this.inFlight.set(dedupKey, promise);
		return promise;
	}

	/**
	 * Register a transient File or ArrayBuffer under a synthetic
	 * `blob:nkido:<uuid>` URI. Returns the URI. Caller MUST call
	 * `unregisterBlob` (e.g. in a try/finally) to free the reference.
	 */
	registerBlob(source: File | ArrayBuffer): string {
		const uuid =
			typeof crypto !== 'undefined' && 'randomUUID' in crypto
				? crypto.randomUUID()
				: Math.random().toString(36).slice(2) + Date.now().toString(36);
		const uri = BLOB_PREFIX + uuid;
		this.blobs.set(uri, source);
		return uri;
	}

	unregisterBlob(uri: string): void {
		this.blobs.delete(uri);
	}

	/** Internal lookup used by the blob handler. */
	resolveBlob(uri: string): File | ArrayBuffer | undefined {
		return this.blobs.get(uri);
	}
}

export const uriResolver = new UriResolver();

export { extractScheme };

File: cedar/src/io/uri_resolver.cpp
#include "cedar/io/uri_resolver.hpp"

#include <cctype>
#include <string>

namespace cedar {

namespace {

bool is_windows_drive_letter(std::string_view uri) {
    return uri.size() >= 2
        && std::isalpha(static_cast<unsigned char>(uri[0]))
        && uri[1] == ':';
}

bool is_bare_path(std::string_view uri) {
    if (uri.empty()) return false;
    if (uri[0] == '/' || uri[0] == '.') return true;
    if (is_windows_drive_letter(uri)) return true;
    return false;
}

std::string to_lower(std::string_view s) {
    std::string out;
    out.reserve(s.size());
    for (char c : s) {
        out.push_back(static_cast<char>(std::tolower(static_cast<unsigned char>(c))));
    }
    return out;
}

}  // namespace

std::string_view UriResolver::extract_scheme(std::string_view uri) {
    if (is_bare_path(uri)) return {};
    auto colon = uri.find(':');
    if (colon == std::string_view::npos) return {};
    return uri.substr(0, colon);
}

UriResolver& UriResolver::instance() {
    static UriResolver resolver;
    return resolver;
}

void UriResolver::register_handler(std::unique_ptr<UriHandler> handler) {
    if (!handler) return;
    std::string key = to_lower(handler->scheme());
    handlers_[std::move(key)] = std::move(handler);
}

const UriHandler* UriResolver::handler_for(std::string_view scheme) const {
    auto it = handlers_.find(to_lower(scheme));
    return it == handlers_.end() ? nullptr : it->second.get();
}

LoadResult UriResolver::load(std::string_view uri) const {
    if (uri.empty()) {
        return {FileLoadError{FileError::InvalidFormat, "empty URI"}};
    }

    std::string_view scheme = extract_scheme(uri);
    if (scheme.empty()) {
        // Bare path — treat as file://
        scheme = "file";
    }

    const UriHandler* h = handler_for(scheme);
    if (!h) {
        return {FileLoadError{FileError::UnsupportedFormat,
                              "no handler registered for scheme '" + std::string(scheme) + "'"}};
    }

    return h->load(uri);
}

}  // namespace cedar
